File: leetcode/cpp/232. 用栈实现队列.cpp
class MyQueue {
public:
    /** Initialize your data structure here. */
    stack<int> s1, s2;
    MyQueue() {

    }

    // 把非空的s1元素push到已空的s2
    void push_s1_s2() {
        if (s2.empty()) {
            while (!s1.empty()) {
                s2.push(s1.top());
                s1.pop();
            }
        }
    }
    
    /** Push element x to the back of queue. */
    void push(int x) {
        s1.push(x);
    }
    
    /** Removes the element from in front of queue and returns that element. */
    int pop() {
        push_s1_s2();
        if (!s2.empty()) {
            int ret = s2.top();
            s2.pop();
            return ret;
        }
        return -1;
    }
    
    /** Get the front element. */
    int peek() {
        push_s1_s2();
        if (!s2.empty()) {
            return s2.top();
        }
        return -1;
    }
    
    /** Returns whether the queue is empty. */
    bool empty() {
        return s1.empty() && s2.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */

File: leetcode/python/232. 用栈实现队列.py
#!/bin/bash python
"""
232. 用栈实现队列

思路：
需要一个额外的栈来完成翻转
"""

class MyQueue(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.s1 = []
        self.s2 = []
    

    def pushOneToTwo(self):
        """
        如果s2为空，则将s1中所有的元素pop/push到s2
        """
        if self.s2 == []:
            while self.s1:
                self.s2.append(self.s1.pop())
        return


    def push(self, x):
        """
        Push element x to the back of queue.
        :type x: int
        :rtype: None
        """
        self.s1.append(x)


    def pop(self):
        """
        Removes the element from in front of queue and returns that element.
        :rtype: int
        """
        self.pushOneToTwo()
        return self.s2.pop() if self.s2 else -1


    def peek(self):
        """
        Get the front element.
        :rtype: int
        """
        self.pushOneToTwo()
        return self.s2[-1] if self.s2 else -1


    def empty(self):
        """
        Returns whether the queue is empty.
        :rtype: bool
        """
        return len(self.s1) == 0 and len(self.s2) == 0


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()